File: weather_forecast.r
# Install and load required packages
if (!require(pacman)) install.packages("pacman")
pacman::p_load(httr, jsonlite, ggplot2, gridExtra, lubridate)

# OpenWeather API key, city, and state
API_KEY <- "58cef9659cdf58d5e41f169d59e058dd"
CITY <- "New York"
STATE <- "NY"

# Function to get weather data
get_weather_data <- function(city, state, api_key) {
  base_url <- "http://api.openweathermap.org/data/2.5/forecast"
  response <- GET(url = base_url,
                  query = list(q = paste(city, state, "US", sep = ","),
                               appid = api_key,
                               units = "metric"))
  
  content <- content(response, "text")
  weather_data <- fromJSON(content)
  return(weather_data)
}

# Get weather data
weather_data <- get_weather_data(CITY, STATE, API_KEY)

# Process data
forecast <- weather_data$list[seq(1, 40, by = 8), ]  # Every 24 hours
forecast$dt <- as.POSIXct(forecast$dt, origin = "1970-01-01")
forecast$date <- format(forecast$dt, "%Y-%m-%d")
forecast$temp <- forecast$main$temp
forecast$description <- sapply(forecast$weather, function(x) x$description[1])

# Create temperature plot
temp_plot <- ggplot(forecast, aes(x = date, y = temp, group = 1)) +
  geom_line(color = "red", size = 1.2) +
  geom_point(color = "red", size = 3) +
  theme_minimal() +
  labs(title = "5-Day Temperature Forecast",
       x = "Date", y = "Temperature (°C)") +
  theme(axis.text.x = element_text(angle = 45, hjust = 1))

# Create weather description plot
desc_plot <- ggplot(forecast, aes(x = date, y = 1, fill = description)) +
  geom_tile() +
  scale_fill_brewer(palette = "Set3") +
  theme_minimal() +
  labs(title = "Weather Description",
       x = "Date", y = NULL, fill = "Description") +
  theme(axis.text.y = element_blank(),
        axis.ticks.y = element_blank(),
        axis.text.x = element_text(angle = 45, hjust = 1))

# Combine plots
infographic <- grid.arrange(
  temp_plot, desc_plot,
  ncol = 1,
  top = textGrob(paste("Weather Forecast for", CITY, STATE),
                 gp = gpar(fontsize = 16, fontface = "bold"))
)

# Save the infographic
ggsave("weather_infographic.png", infographic, width = 10, height = 8, dpi = 300)

print("Weather infographic has been generated and saved as 'weather_infographic.png'.")

File: weather_forecast.py
import requests
import json
from datetime import datetime
from reportlab.lib import colors
from reportlab.lib.pagesizes import letter
from reportlab.platypus import SimpleDocTemplate, Table, TableStyle, Paragraph
from reportlab.lib.styles import getSampleStyleSheet

# OpenWeather API key, city, and state
API_KEY = "58cef9659cdf58d5e41f169d59e058dd"
CITY = "New York"
STATE = "NY"  # Add the state code here

def get_weather_data():
    base_url = "http://api.openweathermap.org/data/2.5/forecast"
    params = {
        "q": f"{CITY},{STATE},US",  # Include state and country code
        "appid": API_KEY,
        "units": "metric"
    }
    response = requests.get(base_url, params=params)
    return json.loads(response.text)

def create_pdf(data):
    doc = SimpleDocTemplate("weather_forecast.pdf", pagesize=letter)
    elements = []
    
    styles = getSampleStyleSheet()
    title = Paragraph(f"5-Day Weather Forecast for {CITY}, {STATE}", styles['Title'])  # Include state in title
    elements.append(title)
    
    table_data = [['Date', 'Time', 'Temperature (°C)', 'Description']]
    for item in data['list'][:40:8]:  # Get data for every 24 hours (3-hour steps)
        date = datetime.fromtimestamp(item['dt'])
        table_data.append([
            date.strftime("%Y-%m-%d"),
            date.strftime("%H:%M"),
            f"{item['main']['temp']:.1f}",
            item['weather'][0]['description']
        ])
    
    table = Table(table_data)
    table.setStyle(TableStyle([
        ('BACKGROUND', (0, 0), (-1, 0), colors.grey),
        ('TEXTCOLOR', (0, 0), (-1, 0), colors.whitesmoke),
        ('ALIGN', (0, 0), (-1, -1), 'CENTER'),
        ('FONTNAME', (0, 0), (-1, 0), 'Helvetica-Bold'),
        ('FONTSIZE', (0, 0), (-1, 0), 14),
        ('BOTTOMPADDING', (0, 0), (-1, 0), 12),
        ('BACKGROUND', (0, 1), (-1, -1), colors.beige),
        ('TEXTCOLOR', (0, 1), (-1, -1), colors.black),
        ('ALIGN', (0, 0), (-1, -1), 'CENTER'),
        ('FONTNAME', (0, 1), (-1, -1), 'Helvetica'),
        ('FONTSIZE', (0, 1), (-1, -1), 12),
        ('TOPPADDING', (0, 1), (-1, -1), 6),
        ('BOTTOMPADDING', (0, 1), (-1, -1), 6),
        ('GRID', (0, 0), (-1, -1), 1, colors.black)
    ]))
    elements.append(table)
    
    doc.build(elements)

def main():
    weather_data = get_weather_data()
    create_pdf(weather_data)
    print(f"Weather forecast PDF for {CITY}, {STATE} has been generated.")

if __name__ == "__main__":
    main()